File: ts/kitstart/src/core/seo/ld.ts
import { ldCompact, localBusiness, type JsonLdNode } from "@evinvest/marketing";
import { freshRating } from "../place/rating";
import { servedLocalities, storefrontOf, type Place } from "../place/types";
import { placeOrigin, siteOrigin, type PlaceView } from "../place/view";
import { contactOf, type Site } from "../site";

/**
 * schema.org, derived from the site, the place and the page's copy — never
 * authored. One `@graph` per page with shared `@id`s, so Google collapses the
 * entities instead of reading each as separate. Output is pinned against
 * aquafix's goldens in `tests/fixtures/kitstart/json-ld.json`, which the Rust
 * `kitstart` feature reads too.
 *
 * A storefront carries its address, pin and photo; a service-area business
 * has none of the three in its type, so its node says where it goes
 * (`areaServed`) and nothing about where it is. No `aggregateRating` from
 * anything baked: only a live rating inside the API's 30-day window.
 */

/** One row of a price list, as the page's table prints it. */
export interface OfferInput {
  name: string;
  /** Whole currency units, the same integer the table cell renders. */
  price: number;
  /** ISO 4217; `EUR` when left out. */
  currency?: string;
}

export interface QuestionAnswer {
  q: string;
  a: string;
}

/** The words one page's graph quotes. */
export interface PageGraphCopy {
  /** The short place name the copy uses: "Royat". */
  placeName: string;
  title: string;
  description: string;
  /** Emitted only where the page renders the price list. */
  offers?: readonly OfferInput[];
  /** Emitted only where the page renders the FAQ. */
  faq?: readonly QuestionAnswer[];
}

const origin = (o: string | null): string => o ?? "";

export const organizationId = <L extends string, P extends string>(site: Site<L, P>): string => `${origin(siteOrigin(site))}/#organization`;
// Language-free: the French and the English page describe one business.
export const businessId = <L extends string, P extends string>(site: Site<L, P>, slug: string): string => `${origin(placeOrigin(site, slug))}/#business`;
export const websiteId = <L extends string, P extends string>(site: Site<L, P>, slug: string): string => `${origin(placeOrigin(site, slug))}/#website`;

export function organizationNode<L extends string, P extends string>(site: Site<L, P>): JsonLdNode {
  return ldCompact({
    "@type": "Organization",
    "@id": organizationId(site),
    name: site.brand.name,
    legalName: site.brand.legalName,
    url: siteOrigin(site),
    email: site.brand.email,
  });
}

function days(place: Place<string>): JsonLdNode[] | undefined {
  return place.hours?.map(h => ({
    "@type": "OpeningHoursSpecification",
    dayOfWeek: h.days,
    opens: h.opens,
    closes: h.closes,
  }));
}

/** Named communes as `Place`s, a radius as a `GeoCircle`. */
export function areaServedNodes(place: Place<string>): JsonLdNode[] {
  const named: JsonLdNode[] = servedLocalities(place).map(name => ({ "@type": "Place", name }));
  const circles: JsonLdNode[] = (place.serviceArea ?? []).flatMap(area =>
    area.kind === "radius"
      ? [
          {
            "@type": "GeoCircle",
            geoMidpoint: { "@type": "GeoCoordinates", latitude: area.center.lat, longitude: area.center.lng },
            geoRadius: area.km * 1000,
          },
        ]
      : [],
  );
  return [...named, ...circles];
}

/** The business behind one place. */
export function businessNode<L extends string, P extends string>(site: Site<L, P>, view: PlaceView<L>, now: Date): JsonLdNode {
  const { place } = view;
  const front = storefrontOf(place);
  const rating = freshRating(place, now);
  return localBusiness(
    {
      id: businessId(site, place.slug),
      type: site.brand.businessType,
      name: place.gbpName,
      url: view.url(site.pages.home),
      telephone: contactOf(site, place).phone ?? undefined,
      email: site.brand.email ?? undefined,
      image: front?.storefrontPhoto ?? undefined,
      address: front
        ? {
            streetAddress: front.address.street,
            postalCode: front.address.postalCode,
            addressLocality: front.address.locality,
            addressRegion: front.address.region,
            addressCountry: front.address.country,
          }
        : undefined,
      geo: front?.geo ?? undefined,
      parentOrganization: { "@id": organizationId(site) },
    },
    ldCompact({
      priceRange: site.brand.priceRange,
      areaServed: areaServedNodes(place),
      openingHoursSpecification: days(place),
      aggregateRating: rating
        ? { "@type": "AggregateRating", ratingValue: rating.value, reviewCount: rating.count, bestRating: 5 }
        : undefined,
    }),
  );
}

/** The town an offer is priced for: a storefront's own, or the first commune served. */
function cityOf(place: Place<string>): JsonLdNode | undefined {
  const name = storefrontOf(place)?.address.locality ?? servedLocalities(place)[0];
  return name === undefined ? undefined : { "@type": "City", name };
}

/** One `Service` + `Offer` per price row: the same integer as the table cell, tax included. */
export function offerNodes<L extends string, P extends string>(site: Site<L, P>, place: Place<L>, offers: readonly OfferInput[]): JsonLdNode[] {
  return offers.map(row => {
    const currency = row.currency ?? "EUR";
    return ldCompact({
      "@type": "Service",
      name: row.name,
      serviceType: row.name,
      provider: { "@id": businessId(site, place.slug) },
      areaServed: cityOf(place),
      offers: {
        "@type": "Offer",
        price: row.price,
        priceCurrency: currency,
        availability: "https://schema.org/InStock",
        priceSpecification: {
          "@type": "PriceSpecification",
          minPrice: row.price,
          priceCurrency: currency,
          valueAddedTaxIncluded: true,
        },
      },
    });
  });
}

export function faqPageNode(items: readonly QuestionAnswer[]): JsonLdNode {
  return {
    "@type": "FAQPage",
    mainEntity: items.map(item => ({
      "@type": "Question",
      name: item.q,
      acceptedAnswer: { "@type": "Answer", text: item.a },
    })),
  };
}

/** Home → page. The chain is the route, so a new page cannot forget it. */
export function breadcrumbNode<L extends string, P extends string>(
  site: Site<L, P>,
  view: PlaceView<L>,
  page: P | "home",
  copy: Pick<PageGraphCopy, "placeName" | "title">,
): JsonLdNode {
  const items: JsonLdNode[] = [
    { "@type": "ListItem", position: 1, name: `${site.brand.name} ${copy.placeName}`, item: view.url(site.pages.home) },
  ];
  if (page !== "home") {
    items.push({ "@type": "ListItem", position: 2, name: copy.title, item: view.url(site.pages[page]) });
  }
  return { "@type": "BreadcrumbList", itemListElement: items };
}

/** The `@graph` for one of a place's pages in one language. */
export function placeGraph<L extends string, P extends string>(
  site: Site<L, P>,
  view: PlaceView<L>,
  page: P | "home",
  copy: PageGraphCopy,
  now: Date,
): JsonLdNode {
  const { place } = view;
  const url = view.url(site.pages[page]);
  const nodes: JsonLdNode[] = [
    organizationNode(site),
    businessNode(site, view, now),
    {
      "@type": "WebSite",
      "@id": websiteId(site, place.slug),
      url: origin(placeOrigin(site, place.slug)),
      name: `${site.brand.name} ${copy.placeName}`,
    },
    {
      "@type": "WebPage",
      "@id": `${url}#page`,
      url,
      name: copy.title,
      description: copy.description,
      inLanguage: site.i18n.hreflangOf(view.locale),
      isPartOf: { "@id": websiteId(site, place.slug) },
      about: { "@id": businessId(site, place.slug) },
    },
    breadcrumbNode(site, view, page, copy),
  ];
  if (copy.offers && copy.offers.length > 0) nodes.push(...offerNodes(site, place, copy.offers));
  if (copy.faq && copy.faq.length > 0) nodes.push(faqPageNode(copy.faq));
  return { "@context": "https://schema.org", "@graph": nodes };
}

File: rust/src/kitstart/ld.rs
//! schema.org, derived from the site, the place and the page's copy — never
//! authored. One `@graph` per page with shared `@id`s, so Google collapses the
//! entities instead of reading each as separate. Pinned, with the TypeScript
//! port, against aquafix's goldens in `tests/fixtures/kitstart/json-ld.json`.
//!
//! A storefront carries its address, pin and photo; a service-area business
//! has none of the three in its type, so its node says where it goes
//! (`areaServed`) and nothing about where it is. No `aggregateRating` from
//! anything baked: only a live rating inside the API's 30-day window.

use jiff::Timestamp;

use super::{
	json::{Json, Object},
	place::{Place, ServiceArea},
	site::{Page, PlaceView, Site},
};

/// One row of a price list, as the page's table prints it.
#[derive(Clone, Debug, PartialEq)]
pub struct OfferInput {
	pub name: String,
	/// Whole currency units, the same number the table cell renders.
	pub price: f64,
	/// ISO 4217; `EUR` when `None`.
	pub currency: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct QuestionAnswer {
	pub q: String,
	pub a: String,
}

/// The words one page's graph quotes.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct PageGraphCopy {
	/// The short place name the copy uses: "Royat".
	pub place_name: String,
	pub title: String,
	pub description: String,
	/// Emitted only where the page renders the price list.
	pub offers: Vec<OfferInput>,
	/// Emitted only where the page renders the FAQ.
	pub faq: Vec<QuestionAnswer>,
}

pub fn organization_id(site: &Site) -> String {
	format!("{}/#organization", site.origin().unwrap_or_default())
}

// Locale-free: the French and the English page describe one business.
pub fn business_id(site: &Site, slug: &str) -> String {
	format!("{}/#business", site.place_origin(slug).unwrap_or_default())
}

pub fn website_id(site: &Site, slug: &str) -> String {
	format!("{}/#website", site.place_origin(slug).unwrap_or_default())
}

pub fn organization_node(site: &Site) -> Object {
	let brand = site.brand();
	Object::typed("Organization")
		.with("@id", organization_id(site))
		.with("name", &brand.name)
		.with("legalName", &brand.legal_name)
		.with("url", site.origin())
		.with("email", brand.email.as_deref())
		.compact()
}

/// Named communes as `Place`s, a radius as a `GeoCircle`.
pub fn area_served_nodes(place: &Place) -> Vec<Object> {
	let named = place.served_localities().into_iter().map(|name| Object::typed("Place").with("name", name));
	let circles = place.service_area.iter().flatten().filter_map(|area| match area {
		ServiceArea::Radius { center, km } => Some(
			Object::typed("GeoCircle")
				.with("geoMidpoint", Object::typed("GeoCoordinates").with("latitude", center.lat).with("longitude", center.lng))
				.with("geoRadius", km * 1000.0),
		),
		ServiceArea::Localities(_) => None,
	});
	named.chain(circles).collect()
}

fn opening_hours(place: &Place) -> Option<Vec<Object>> {
	place.hours.as_ref().map(|hours| {
		hours
			.iter()
			.map(|h| {
				Object::typed("OpeningHoursSpecification")
					.with("dayOfWeek", h.days.iter().map(|d| Json::from(d.as_str())).collect::<Vec<_>>())
					.with("opens", &h.opens)
					.with("closes", &h.closes)
			})
			.collect()
	})
}

/// The business behind one place.
pub fn business_node(view: &PlaceView<'_>, now: Timestamp) -> Object {
	let site = view.site;
	let place = view.place;
	let front = place.storefront();
	let address = front.map(|f| {
		Object::typed("PostalAddress")
			.with("streetAddress", &f.address.street)
			.with("postalCode", &f.address.postal_code)
			.with("addressLocality", &f.address.locality)
			.with("addressRegion", &f.address.region)
			.with("addressCountry", &f.address.country)
			.compact()
	});
	let geo = front
		.and_then(|f| f.geo)
		.map(|g| Object::typed("GeoCoordinates").with("latitude", g.lat).with("longitude", g.lng));
	let rating = place.fresh_rating(now).map(|r| {
		Object::typed("AggregateRating")
			.with("ratingValue", r.value)
			.with("reviewCount", r.count)
			.with("bestRating", 5u32)
	});
	// Key order is `@evinvest/marketing`'s `localBusiness`: its own fields,
	// then address and geo, then the caller's extras.
	let extra = Object::new()
		.with("priceRange", site.brand().price_range.as_deref())
		.with("areaServed", area_served_nodes(place))
		.with("openingHoursSpecification", opening_hours(place))
		.with("aggregateRating", rating)
		.compact();
	Object::typed(&site.brand().business_type)
		.with("@id", business_id(site, &place.slug))
		.with("name", &place.gbp_name)
		.with("url", view.url(site.home().suffix))
		.with("telephone", site.contact_of(place).phone)
		.with("email", site.brand().email.as_deref())
		.with("image", front.and_then(|f| f.storefront_photo.as_deref()))
		.with("parentOrganization", Object::reference(organization_id(site)))
		.with("address", address)
		.with("geo", geo)
		.extend(extra)
		.compact()
}

/// The town an offer is priced for: a storefront's own, or the first commune
/// served.
fn city_of(place: &Place) -> Option<Object> {
	let name = place.storefront().map(|f| f.address.locality.as_str()).or_else(|| place.served_localities().first().copied())?;
	Some(Object::typed("City").with("name", name))
}

/// One `Service` + `Offer` per price row: the same number as the table cell,
/// tax included.
pub fn offer_nodes(site: &Site, place: &Place, offers: &[OfferInput]) -> Vec<Object> {
	offers
		.iter()
		.map(|row| {
			let currency = row.currency.as_deref().unwrap_or("EUR");
			Object::typed("Service")
				.with("name", &row.name)
				.with("serviceType", &row.name)
				.with("provider", Object::reference(business_id(site, &place.slug)))
				.with("areaServed", city_of(place))
				.with(
					"offers",
					Object::typed("Offer")
						.with("price", row.price)
						.with("priceCurrency", currency)
						.with("availability", "https://schema.org/InStock")
						.with(
							"priceSpecification",
							Object::typed("PriceSpecification")
								.with("minPrice", row.price)
								.with("priceCurrency", currency)
								.with("valueAddedTaxIncluded", true),
						),
				)
				.compact()
		})
		.collect()
}

pub fn faq_page_node(items: &[QuestionAnswer]) -> Object {
	let questions: Vec<Object> = items
		.iter()
		.map(|item| {
			Object::typed("Question")
				.with("name", &item.q)
				.with("acceptedAnswer", Object::typed("Answer").with("text", &item.a))
		})
		.collect();
	Object::typed("FAQPage").with("mainEntity", questions)
}

/// Home → page. The chain is the route, so a new page cannot forget it.
pub fn breadcrumb_node(view: &PlaceView<'_>, page: Page<'_>, copy: &PageGraphCopy) -> Object {
	let site = view.site;
	let mut items = vec![
		Object::typed("ListItem")
			.with("position", 1u32)
			.with("name", format!("{} {}", site.brand().name, copy.place_name))
			.with("item", view.url(site.home().suffix)),
	];
	if !page.is_home() {
		items.push(Object::typed("ListItem").with("position", 2u32).with("name", &copy.title).with("item", view.url(page.suffix)));
	}
	Object::typed("BreadcrumbList").with("itemListElement", items)
}

/// The `@graph` for one of a place's pages in one locale.
///
/// Render it with [`Json::to_script`] inside `<script type="application/ld+json">`.
pub fn place_graph(view: &PlaceView<'_>, page: Page<'_>, copy: &PageGraphCopy, now: Timestamp) -> Json {
	let site = view.site;
	let slug = view.place.slug.as_str();
	let url = view.url(page.suffix);
	let mut nodes = vec![
		organization_node(site),
		business_node(view, now),
		Object::typed("WebSite")
			.with("@id", website_id(site, slug))
			.with("url", site.place_origin(slug).unwrap_or_default())
			.with("name", format!("{} {}", site.brand().name, copy.place_name)),
		Object::typed("WebPage")
			.with("@id", format!("{url}#page"))
			.with("url", url)
			.with("name", &copy.title)
			.with("description", &copy.description)
			.with("inLanguage", site.i18n().hreflang_of(view.locale))
			.with("isPartOf", Object::reference(website_id(site, slug)))
			.with("about", Object::reference(business_id(site, slug))),
		breadcrumb_node(view, page, copy),
	];
	if !copy.offers.is_empty() {
		nodes.extend(offer_nodes(site, view.place, &copy.offers));
	}
	if !copy.faq.is_empty() {
		nodes.push(faq_page_node(&copy.faq));
	}
	Object::new().with("@context", "https://schema.org").with("@graph", nodes).into()
}
